// ngx-img-cropper/esm2015/lib/image-cropper/imageCropperDataShare.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
export class ImageCropperDataShare {
    constructor() {
        this.share = {};
    }
    /**
     * @param {?} canvas
     * @return {?}
     */
    setPressed(canvas) {
        this.pressed = canvas;
    }
    /**
     * @param {?} canvas
     * @return {?}
     */
    setReleased(canvas) {
        if (canvas === this.pressed) {
            //  this.pressed = undefined;
        }
    }
    /**
     * @param {?} canvas
     * @return {?}
     */
    setOver(canvas) {
        this.over = canvas;
    }
    /**
     * @param {?} canvas
     * @param {?} style
     * @return {?}
     */
    setStyle(canvas, style) {
        if (this.pressed !== undefined) {
            if (this.pressed === canvas) {
                // TODO: check this
                // angular.element(document.documentElement).css('cursor', style);
            }
        }
        else {
            if (canvas === this.over) {
                // TODO: check this
                // angular.element(document.documentElement).css('cursor', style);
            }
        }
    }
}
if (false) {
    /** @type {?} */
    ImageCropperDataShare.prototype.share;
    /** @type {?} */
    ImageCropperDataShare.prototype.pressed;
    /** @type {?} */
    ImageCropperDataShare.prototype.over;
}
