// ngx-img-cropper/esm5/lib/image-cropper/cropper-draw-settings.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var CropperDrawSettings = /** @class */ (function () {
    function CropperDrawSettings(settings) {
        this.lineDash = false;
        this.strokeWidth = 1;
        this.strokeColor = 'rgba(255,255,255,1)';
        this.fillColor = 'rgba(255,255,255,1)';
        this.dragIconStrokeWidth = 1;
        this.dragIconStrokeColor = 'rgba(0,0,0,1)';
        this.dragIconFillColor = 'rgba(255,255,255,1)';
        this.backgroundFillColor = 'rgba(0,0,0,0.6)';
        if (typeof settings === 'object') {
            this.lineDash = settings.lineDash || this.lineDash;
            this.strokeWidth = settings.strokeWidth || this.strokeWidth;
            this.fillColor = settings.fillColor || this.fillColor;
            this.strokeColor = settings.strokeColor || this.strokeColor;
            this.dragIconStrokeWidth =
                settings.dragIconStrokeWidth || this.dragIconStrokeWidth;
            this.dragIconStrokeColor =
                settings.dragIconStrokeColor || this.dragIconStrokeColor;
            this.dragIconFillColor =
                settings.dragIconFillColor || this.dragIconFillColor;
            this.backgroundFillColor =
                settings.backgroundFillColor || this.backgroundFillColor;
        }
    }
    return CropperDrawSettings;
}());
export { CropperDrawSettings };
if (false) {
    /** @type {?} */
    CropperDrawSettings.prototype.lineDash;
    /** @type {?} */
    CropperDrawSettings.prototype.strokeWidth;
    /** @type {?} */
    CropperDrawSettings.prototype.strokeColor;
    /** @type {?} */
    CropperDrawSettings.prototype.fillColor;
    /** @type {?} */
    CropperDrawSettings.prototype.dragIconStrokeWidth;
    /** @type {?} */
    CropperDrawSettings.prototype.dragIconStrokeColor;
    /** @type {?} */
    CropperDrawSettings.prototype.dragIconFillColor;
    /** @type {?} */
    CropperDrawSettings.prototype.backgroundFillColor;
}
