// ngx-img-cropper/esm5/lib/image-cropper/cropper-settings.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { CropperDrawSettings } from './cropper-draw-settings';
/**
 * @record
 */
export function ICropperSettings() { }
if (false) {
    /** @type {?|undefined} */
    ICropperSettings.prototype.canvasWidth;
    /** @type {?|undefined} */
    ICropperSettings.prototype.canvasHeight;
    /** @type {?|undefined} */
    ICropperSettings.prototype.dynamicSizing;
    /** @type {?|undefined} */
    ICropperSettings.prototype.cropperClass;
    /** @type {?|undefined} */
    ICropperSettings.prototype.croppingClass;
    /** @type {?|undefined} */
    ICropperSettings.prototype.width;
    /** @type {?|undefined} */
    ICropperSettings.prototype.height;
    /** @type {?|undefined} */
    ICropperSettings.prototype.minWidth;
    /** @type {?|undefined} */
    ICropperSettings.prototype.minHeight;
    /** @type {?|undefined} */
    ICropperSettings.prototype.minWithRelativeToResolution;
    /** @type {?|undefined} */
    ICropperSettings.prototype.croppedWidth;
    /** @type {?|undefined} */
    ICropperSettings.prototype.croppedHeight;
    /** @type {?|undefined} */
    ICropperSettings.prototype.cropperDrawSettings;
    /** @type {?|undefined} */
    ICropperSettings.prototype.touchRadius;
    /** @type {?|undefined} */
    ICropperSettings.prototype.noFileInput;
    /** @type {?|undefined} */
    ICropperSettings.prototype.fileType;
    /** @type {?|undefined} */
    ICropperSettings.prototype.resampleFn;
    /** @type {?|undefined} */
    ICropperSettings.prototype.markerSizeMultiplier;
    /** @type {?|undefined} */
    ICropperSettings.prototype.centerTouchRadius;
    /** @type {?|undefined} */
    ICropperSettings.prototype.showCenterMarker;
    /** @type {?|undefined} */
    ICropperSettings.prototype.allowedFilesRegex;
    /** @type {?|undefined} */
    ICropperSettings.prototype.cropOnResize;
    /** @type {?|undefined} */
    ICropperSettings.prototype.preserveSize;
    /** @type {?|undefined} */
    ICropperSettings.prototype.compressRatio;
    /** @type {?|undefined} */
    ICropperSettings.prototype.rounded;
    /** @type {?|undefined} */
    ICropperSettings.prototype.keepAspect;
}
var CropperSettings = /** @class */ (function () {
    function CropperSettings(settings) {
        this.canvasWidth = 300;
        this.canvasHeight = 300;
        this.dynamicSizing = false;
        this.width = 200;
        this.height = 200;
        this.minWidth = 50;
        this.minHeight = 50;
        this.minWithRelativeToResolution = true;
        this.croppedWidth = 100;
        this.croppedHeight = 100;
        this.cropperDrawSettings = new CropperDrawSettings();
        this.touchRadius = 20;
        this.noFileInput = false;
        this.markerSizeMultiplier = 1;
        this.centerTouchRadius = 20;
        this.showCenterMarker = true;
        this.allowedFilesRegex = /\.(jpe?g|png|gif|bmp)$/i;
        this.cropOnResize = true;
        this.preserveSize = false;
        this.compressRatio = 1.0;
        // tslint:disable-next-line:variable-name
        this._rounded = false;
        // tslint:disable-next-line:variable-name
        this._keepAspect = true;
        if (typeof settings === 'object') {
            Object.assign(this, settings);
        }
    }
    Object.defineProperty(CropperSettings.prototype, "rounded", {
        get: /**
         * @return {?}
         */
        function () {
            return this._rounded;
        },
        set: /**
         * @param {?} val
         * @return {?}
         */
        function (val) {
            this._rounded = val;
            if (val) {
                this._keepAspect = true;
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(CropperSettings.prototype, "keepAspect", {
        get: /**
         * @return {?}
         */
        function () {
            return this._keepAspect;
        },
        set: /**
         * @param {?} val
         * @return {?}
         */
        function (val) {
            this._keepAspect = val;
            if (this._rounded === true && this._keepAspect === false) {
                console.error('Cannot set keep aspect to false on rounded cropper. Ellipsis not supported');
                this._keepAspect = true;
            }
        },
        enumerable: true,
        configurable: true
    });
    return CropperSettings;
}());
export { CropperSettings };
if (false) {
    /** @type {?} */
    CropperSettings.prototype.canvasWidth;
    /** @type {?} */
    CropperSettings.prototype.canvasHeight;
    /** @type {?} */
    CropperSettings.prototype.dynamicSizing;
    /** @type {?} */
    CropperSettings.prototype.cropperClass;
    /** @type {?} */
    CropperSettings.prototype.croppingClass;
    /** @type {?} */
    CropperSettings.prototype.width;
    /** @type {?} */
    CropperSettings.prototype.height;
    /** @type {?} */
    CropperSettings.prototype.minWidth;
    /** @type {?} */
    CropperSettings.prototype.minHeight;
    /** @type {?} */
    CropperSettings.prototype.minWithRelativeToResolution;
    /** @type {?} */
    CropperSettings.prototype.croppedWidth;
    /** @type {?} */
    CropperSettings.prototype.croppedHeight;
    /** @type {?} */
    CropperSettings.prototype.cropperDrawSettings;
    /** @type {?} */
    CropperSettings.prototype.touchRadius;
    /** @type {?} */
    CropperSettings.prototype.noFileInput;
    /** @type {?} */
    CropperSettings.prototype.fileType;
    /** @type {?} */
    CropperSettings.prototype.resampleFn;
    /** @type {?} */
    CropperSettings.prototype.markerSizeMultiplier;
    /** @type {?} */
    CropperSettings.prototype.centerTouchRadius;
    /** @type {?} */
    CropperSettings.prototype.showCenterMarker;
    /** @type {?} */
    CropperSettings.prototype.allowedFilesRegex;
    /** @type {?} */
    CropperSettings.prototype.cropOnResize;
    /** @type {?} */
    CropperSettings.prototype.preserveSize;
    /** @type {?} */
    CropperSettings.prototype.compressRatio;
    /**
     * @type {?}
     * @private
     */
    CropperSettings.prototype._rounded;
    /**
     * @type {?}
     * @private
     */
    CropperSettings.prototype._keepAspect;
}
