// ngx-img-cropper/esm5/lib/image-cropper/imageCropperDataShare.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var ImageCropperDataShare = /** @class */ (function () {
    function ImageCropperDataShare() {
        this.share = {};
    }
    /**
     * @param {?} canvas
     * @return {?}
     */
    ImageCropperDataShare.prototype.setPressed = /**
     * @param {?} canvas
     * @return {?}
     */
    function (canvas) {
        this.pressed = canvas;
    };
    /**
     * @param {?} canvas
     * @return {?}
     */
    ImageCropperDataShare.prototype.setReleased = /**
     * @param {?} canvas
     * @return {?}
     */
    function (canvas) {
        if (canvas === this.pressed) {
            //  this.pressed = undefined;
        }
    };
    /**
     * @param {?} canvas
     * @return {?}
     */
    ImageCropperDataShare.prototype.setOver = /**
     * @param {?} canvas
     * @return {?}
     */
    function (canvas) {
        this.over = canvas;
    };
    /**
     * @param {?} canvas
     * @param {?} style
     * @return {?}
     */
    ImageCropperDataShare.prototype.setStyle = /**
     * @param {?} canvas
     * @param {?} style
     * @return {?}
     */
    function (canvas, style) {
        if (this.pressed !== undefined) {
            if (this.pressed === canvas) {
                // TODO: check this
                // angular.element(document.documentElement).css('cursor', style);
            }
        }
        else {
            if (canvas === this.over) {
                // TODO: check this
                // angular.element(document.documentElement).css('cursor', style);
            }
        }
    };
    return ImageCropperDataShare;
}());
export { ImageCropperDataShare };
if (false) {
    /** @type {?} */
    ImageCropperDataShare.prototype.share;
    /** @type {?} */
    ImageCropperDataShare.prototype.pressed;
    /** @type {?} */
    ImageCropperDataShare.prototype.over;
}
