// ngx-img-cropper/esm5/lib/image-cropper/model/bounds.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { PointPool } from './pointPool';
var Bounds = /** @class */ (function () {
    function Bounds(x, y, width, height) {
        if (x === void 0) {
            x = 0;
        }
        if (y === void 0) {
            y = 0;
        }
        if (width === void 0) {
            width = 0;
        }
        if (height === void 0) {
            height = 0;
        }
        this.left = x;
        this.right = x + width;
        this.top = y;
        this.bottom = y + height;
    }
    Object.defineProperty(Bounds.prototype, "width", {
        get: /**
         * @return {?}
         */
        function () {
            return this.right - this.left;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Bounds.prototype, "height", {
        get: /**
         * @return {?}
         */
        function () {
            return this.bottom - this.top;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    Bounds.prototype.getCentre = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var w = this.width;
        /** @type {?} */
        var h = this.height;
        return new PointPool().instance.borrow(this.left + w / 2, this.top + h / 2);
    };
    return Bounds;
}());
export { Bounds };
if (false) {
    /** @type {?} */
    Bounds.prototype.left;
    /** @type {?} */
    Bounds.prototype.right;
    /** @type {?} */
    Bounds.prototype.top;
    /** @type {?} */
    Bounds.prototype.bottom;
}
