// ngx-img-cropper/esm5/lib/image-cropper/model/cropPosition.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Bounds } from './bounds';
var CropPosition = /** @class */ (function () {
    function CropPosition(x, y, w, h) {
        if (x === void 0) { x = 0; }
        if (y === void 0) { y = 0; }
        if (w === void 0) { w = 0; }
        if (h === void 0) { h = 0; }
        this.x = +x;
        this.y = +y;
        this.w = +w;
        this.h = +h;
    }
    /**
     * @return {?}
     */
    CropPosition.prototype.toBounds = /**
     * @return {?}
     */
    function () {
        return new Bounds(this.x, this.y, this.w, this.h);
    };
    /**
     * @return {?}
     */
    CropPosition.prototype.isInitialized = /**
     * @return {?}
     */
    function () {
        return this.x !== 0 && this.y !== 0 && this.w !== 0 && this.h !== 0;
    };
    return CropPosition;
}());
export { CropPosition };
if (false) {
    /** @type {?} */
    CropPosition.prototype.x;
    /** @type {?} */
    CropPosition.prototype.y;
    /** @type {?} */
    CropPosition.prototype.w;
    /** @type {?} */
    CropPosition.prototype.h;
}
