// ngx-img-cropper/esm5/lib/image-cropper/model/point.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @record
 */
export function IPoint() { }
if (false) {
    /** @type {?} */
    IPoint.prototype.x;
    /** @type {?} */
    IPoint.prototype.y;
    /** @type {?} */
    IPoint.prototype.next;
    /** @type {?} */
    IPoint.prototype.prev;
}
var Point = /** @class */ (function () {
    function Point(x, y) {
        if (x === void 0) { x = 0; }
        if (y === void 0) { y = 0; }
        this.x = x;
        this.y = y;
    }
    Object.defineProperty(Point.prototype, "next", {
        get: /**
         * @return {?}
         */
        function () {
            return this.myNext;
        },
        set: /**
         * @param {?} p
         * @return {?}
         */
        function (p) {
            this.myNext = p;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Point.prototype, "prev", {
        get: /**
         * @return {?}
         */
        function () {
            return this.myPrev;
        },
        set: /**
         * @param {?} p
         * @return {?}
         */
        function (p) {
            this.myPrev = p;
        },
        enumerable: true,
        configurable: true
    });
    return Point;
}());
export { Point };
if (false) {
    /** @type {?} */
    Point.prototype.x;
    /** @type {?} */
    Point.prototype.y;
    /**
     * @type {?}
     * @private
     */
    Point.prototype.myNext;
    /**
     * @type {?}
     * @private
     */
    Point.prototype.myPrev;
}
