// ngx-img-cropper/esm2015/lib/image-cropper/model/bounds.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { PointPool } from './pointPool';
export class Bounds {
    /**
     * @param {?=} x
     * @param {?=} y
     * @param {?=} width
     * @param {?=} height
     */
    constructor(x, y, width, height) {
        if (x === void 0) {
            x = 0;
        }
        if (y === void 0) {
            y = 0;
        }
        if (width === void 0) {
            width = 0;
        }
        if (height === void 0) {
            height = 0;
        }
        this.left = x;
        this.right = x + width;
        this.top = y;
        this.bottom = y + height;
    }
    /**
     * @return {?}
     */
    get width() {
        return this.right - this.left;
    }
    /**
     * @return {?}
     */
    get height() {
        return this.bottom - this.top;
    }
    /**
     * @return {?}
     */
    getCentre() {
        /** @type {?} */
        const w = this.width;
        /** @type {?} */
        const h = this.height;
        return new PointPool().instance.borrow(this.left + w / 2, this.top + h / 2);
    }
}
if (false) {
    /** @type {?} */
    Bounds.prototype.left;
    /** @type {?} */
    Bounds.prototype.right;
    /** @type {?} */
    Bounds.prototype.top;
    /** @type {?} */
    Bounds.prototype.bottom;
}
