// ngx-img-cropper/esm2015/lib/image-cropper/model/handle.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Point } from './point';
import { CropperSettings } from '../cropper-settings';
/**
 * @record
 */
export function IHandle() { }
if (false) {
    /** @type {?} */
    IHandle.prototype.over;
    /** @type {?} */
    IHandle.prototype.drag;
    /** @type {?} */
    IHandle.prototype.position;
    /** @type {?} */
    IHandle.prototype.offset;
    /** @type {?} */
    IHandle.prototype.radius;
    /** @type {?} */
    IHandle.prototype.cropperSettings;
    /**
     * @param {?} x
     * @param {?} y
     * @return {?}
     */
    IHandle.prototype.setPosition = function (x, y) { };
    /**
     * @param {?} value
     * @return {?}
     */
    IHandle.prototype.setDrag = function (value) { };
    /**
     * @param {?} ctx
     * @return {?}
     */
    IHandle.prototype.draw = function (ctx) { };
    /**
     * @param {?} over
     * @return {?}
     */
    IHandle.prototype.setOver = function (over) { };
    /**
     * @param {?} x
     * @param {?} y
     * @return {?}
     */
    IHandle.prototype.touchInBounds = function (x, y) { };
}
export class Handle {
    /**
     * @param {?} x
     * @param {?} y
     * @param {?} radius
     * @param {?} settings
     */
    constructor(x, y, radius, settings) {
        this.cropperSettings = new CropperSettings();
        this.over = false;
        this.drag = false;
        this._position = new Point(x, y);
        this.offset = new Point(0, 0);
        this.radius = radius;
        this.cropperSettings = settings;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    setDrag(value) {
        this.drag = value;
        this.setOver(value);
    }
    /**
     * @param {?} ctx
     * @return {?}
     */
    draw(ctx) {
        // this should't be empty
    }
    /**
     * @param {?} over
     * @return {?}
     */
    setOver(over) {
        this.over = over;
    }
    /**
     * @param {?} x
     * @param {?} y
     * @return {?}
     */
    touchInBounds(x, y) {
        return (x > this.position.x - this.radius + this.offset.x &&
            x < this.position.x + this.radius + this.offset.x &&
            y > this.position.y - this.radius + this.offset.y &&
            y < this.position.y + this.radius + this.offset.y);
    }
    /**
     * @return {?}
     */
    get position() {
        return this._position;
    }
    /**
     * @param {?} x
     * @param {?} y
     * @return {?}
     */
    setPosition(x, y) {
        this._position.x = x;
        this._position.y = y;
    }
}
if (false) {
    /** @type {?} */
    Handle.prototype.over;
    /** @type {?} */
    Handle.prototype.drag;
    /**
     * @type {?}
     * @private
     */
    Handle.prototype._position;
    /** @type {?} */
    Handle.prototype.offset;
    /** @type {?} */
    Handle.prototype.radius;
    /** @type {?} */
    Handle.prototype.cropperSettings;
}
