// ngx-img-cropper/esm2015/lib/image-cropper/model/point.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @record
 */
export function IPoint() { }
if (false) {
    /** @type {?} */
    IPoint.prototype.x;
    /** @type {?} */
    IPoint.prototype.y;
    /** @type {?} */
    IPoint.prototype.next;
    /** @type {?} */
    IPoint.prototype.prev;
}
export class Point {
    /**
     * @param {?=} x
     * @param {?=} y
     */
    constructor(x = 0, y = 0) {
        this.x = x;
        this.y = y;
    }
    /**
     * @return {?}
     */
    get next() {
        return this.myNext;
    }
    /**
     * @param {?} p
     * @return {?}
     */
    set next(p) {
        this.myNext = p;
    }
    /**
     * @return {?}
     */
    get prev() {
        return this.myPrev;
    }
    /**
     * @param {?} p
     * @return {?}
     */
    set prev(p) {
        this.myPrev = p;
    }
}
if (false) {
    /** @type {?} */
    Point.prototype.x;
    /** @type {?} */
    Point.prototype.y;
    /**
     * @type {?}
     * @private
     */
    Point.prototype.myNext;
    /**
     * @type {?}
     * @private
     */
    Point.prototype.myPrev;
}
